### DISCUSSION TITLE
Using with Wikidata

### DISCUSSION THREAD

**Discussion by nyurik** [oxigraph/oxigraph#202] [General]
I would love for Oxigraph to become the primary engine for Wikidata. Yet, the [latest report](https://m.wikidata.org/wiki/Wikidata:SPARQL_query_service/WDQS_backend_update/WDQS_backend_alternatives), with an in-depth [paper](https://m.wikidata.org/wiki/File:WDQS_Backend_Alternatives_working_paper.pdf) mentions OxiGraph as an **Early in development (Version 0.3); Targeted as embedded store; Identified as "hobby project"** -- without mentioning what exactly is missing feature-wise and/or stability-wise.   I would like to start a discussion comparing OxiGraph with other projects identified by that research, and eventually be able to exceed other project capabilities.

**Comment by Tpt** [comment]
Thank you for starting this discussion!

Imho, it makes sense for the WMF to not include Oxigraph in the "new system for Wikidata Query Service" project. They do not want to have to commit again to maintain again an abandoned project. Oxigraph is still a "one guy hobby project" so with a very significant likelihood of being abandoned.

To give my wet finger beliefs on Oxigraph vs the identified projects and Blazegraph:
* Jena, RDF4J and Virtuoso are old stable projects (~10 years) with active development funded by external companies. QLever is a recent research project so might get abandoned and Oxigraph is a one guy hobby.
* Virtuoso and QLever do not implement full SPARQL 1.1. The other systems do.
* At my knowledge (but I might be wrong) RDF4J and Jena storage backends have still a global lock for write so are likely to encounter the same write contention issues as Blazegraph. Oxigraph and I believe Virtuoso supports multiple writers and QLevel is currently read only. Jena main developer [is currently experimenting with a new storage backend](https://github.com/afs/TDB3) based also on RocksDB that might solve that.
* Oxigraph query planner/optimizer is currently very primitive and gets much slower queries than Blazegraph. I have not done benchmarks on complex queries between the compared system. I would tend to believe that Virtuoso is likely to be significantly faster than Jena/RDF4J but I don't find recent benchmarks. QLever is likely to be the fastest [according to their paper](https://ad-publications.cs.uni-freiburg.de/CIKM_qlever_BB_2017.pdf) but QLever is still read-only and it's much easier to make very fast read-only database.

**Comment by afs** [comment]
Good analysis.

> active development funded by external companies

Jena gets some contributions from external companies sometimes.

**Reply by Tpt** [reply]
> Jena gets some contributions from external companies sometimes.

I was not aware of that. It makes Jena look even more impressive! Congrats!
